File: test-license.js
// SocialTouch License Test Script
// API通信テスト用JavaScriptスクリプト

const API_BASE_URL = "http://localhost:3001/api";
const TEST_DEVICE_HASH = "58ff07d6539b1b8c";

// ライセンス認証テスト
async function testLicenseVerification() {
    console.log("=== ライセンス認証テスト開始 ===");

    try {
        const response = await fetch(`${API_BASE_URL}/license/verify`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                device_hash: TEST_DEVICE_HASH
            })
        });

        const data = await response.json();

        console.log("📊 認証結果:");
        console.log("  - success:", data.success);
        console.log("  - is_valid:", data.is_valid);
        console.log("  - status:", data.status);
        console.log("  - expires_at:", data.expires_at);
        console.log("  - cached:", data.cached);

        if (data.success && data.is_valid) {
            console.log("✅ ライセンス認証成功！");
            return { success: true, expires_at: data.expires_at };
        } else {
            console.log("❌ ライセンス認証失敗");
            return { success: false };
        }
    } catch (error) {
        console.error("❌ API通信エラー:", error.message);
        return { success: false, error: error.message };
    }
}

// デバイスハッシュ生成シミュレーション
function simulateDeviceHashGeneration() {
    console.log("=== デバイスハッシュ生成テスト ===");

    const deviceId = "test_device_001";
    const model = "iPhone";
    const data = `${deviceId}:${model}:socialtouch`;

    console.log("ハッシュ元データ:", data);

    // 簡易ハッシュ生成（Luaスクリプトと同じロジック）
    let hash = "";
    let sum = 0;

    for (let i = 0; i < data.length; i++) {
        sum += data.charCodeAt(i);
    }

    // Math.randomをseedで初期化する代替手段
    let seed = sum;
    function seededRandom() {
        seed = (seed * 9301 + 49297) % 233280;
        return seed / 233280;
    }

    for (let i = 0; i < 16; i++) {
        const n = Math.floor(seededRandom() * 16);
        if (n < 10) {
            hash += n.toString();
        } else {
            hash += String.fromCharCode(87 + n); // a-f
        }
    }

    console.log("生成されたハッシュ:", hash);
    return hash;
}

// キャッシュ機能テスト
async function testCaching() {
    console.log("=== キャッシュ機能テスト ===");

    console.log("1回目のリクエスト（キャッシュなし）:");
    const result1 = await testLicenseVerification();

    console.log("\n2回目のリクエスト（キャッシュあり）:");
    const result2 = await testLicenseVerification();

    return { result1, result2 };
}

// 期限切れデバイスハッシュのテスト
async function testExpiredDevice() {
    console.log("=== 期限切れデバイステスト ===");

    try {
        const response = await fetch(`${API_BASE_URL}/license/verify`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                device_hash: "expired_device_hash"
            })
        });

        const data = await response.json();

        console.log("期限切れデバイスの結果:");
        console.log("  - success:", data.success);
        console.log("  - is_valid:", data.is_valid);
        console.log("  - error:", data.error);

        return data;
    } catch (error) {
        console.error("API通信エラー:", error.message);
        return { success: false, error: error.message };
    }
}

// API健全性チェック
async function healthCheck() {
    console.log("=== API健全性チェック ===");

    try {
        const response = await fetch(`${API_BASE_URL}/license/verify`, {
            method: 'GET'
        });

        const data = await response.json();
        console.log("Health Check結果:", data);
        return data;
    } catch (error) {
        console.error("Health Check失敗:", error.message);
        return { success: false, error: error.message };
    }
}

// メイン実行
async function main() {
    console.log("==========================================");
    console.log("    SocialTouch License Test Script      ");
    console.log("==========================================\n");

    // 1. デバイスハッシュ生成テスト
    const generatedHash = simulateDeviceHashGeneration();
    console.log("");

    // 2. API健全性チェック
    await healthCheck();
    console.log("");

    // 3. 有効なライセンス認証テスト
    const licenseResult = await testLicenseVerification();
    console.log("");

    // 4. キャッシュ機能テスト
    await testCaching();
    console.log("");

    // 5. 期限切れデバイステスト
    await testExpiredDevice();
    console.log("");

    // 結果サマリー
    console.log("==========================================");
    if (licenseResult.success) {
        console.log("🎉 すべてのテストが正常に完了しました");
        console.log("📅 有効期限:", licenseResult.expires_at);
        console.log("");
        console.log("📱 実際のAutoTouchスクリプトでは以下の処理を実行:");
        console.log("  1. ツール選択画面を表示");
        console.log("  2. 選択されたツールを実行");
        console.log("  3. ライセンス認証は24時間キャッシュされます");
    } else {
        console.log("⚠️ ライセンス認証に失敗しました");
        console.log("📝 実際のスクリプトでは登録画面を表示します");
    }
    console.log("==========================================");
}

// テスト実行
main().catch(console.error);

File: lua/test_license.lua
-- ==========================================
-- SocialTouch ライセンステスト用スクリプト
-- API通信テスト
-- ==========================================

-- ログ出力設定
local function log(message)
    print(os.date("[%Y-%m-%d %H:%M:%S] ") .. message)
end

-- 設定
local Config = {
    API_BASE_URL = "http://localhost:3001/api",
    TEST_DEVICE_HASH = "58ff07d6539b1b8c"
}

-- HTTP リクエスト（シミュレーション）
local function httpRequest(url, method, data)
    log(string.format("🌐 HTTPリクエスト: %s %s", method, url))

    -- 実際のHTTPリクエストの代わりにcurlコマンドを使用
    local jsonData = ""
    if data then
        local parts = {}
        for k, v in pairs(data) do
            table.insert(parts, string.format('"%s":"%s"', k, v))
        end
        jsonData = "{" .. table.concat(parts, ",") .. "}"
    end

    local command = string.format('curl -s -X %s %s -H "Content-Type: application/json" -d \'%s\'',
                                   method, url, jsonData)

    log("実行コマンド: " .. command)

    local handle = io.popen(command)
    local response = handle:read("*a")
    handle:close()

    if not response or response == "" then
        return nil, "HTTPリクエストに失敗しました"
    end

    log("レスポンス: " .. response)

    -- 簡易JSON解析
    local result = {}
    for key, value in string.gmatch(response, '"([%w_]+)":"?([^",}]+)"?') do
        if value == "true" then
            result[key] = true
        elseif value == "false" then
            result[key] = false
        else
            result[key] = value
        end
    end

    return result, nil
end

-- ライセンス認証テスト
local function testLicenseVerification()
    log("=== ライセンス認証テスト開始 ===")

    local url = Config.API_BASE_URL .. "/license/verify"
    local data = { device_hash = Config.TEST_DEVICE_HASH }

    local response, err = httpRequest(url, "POST", data)

    if err then
        log("❌ API通信エラー: " .. err)
        return false
    end

    if not response then
        log("❌ レスポンスが空です")
        return false
    end

    log("📊 認証結果:")
    log("  - success: " .. tostring(response.success))
    log("  - is_valid: " .. tostring(response.is_valid))
    log("  - status: " .. tostring(response.status))
    log("  - expires_at: " .. tostring(response.expires_at))

    if response.success and response.is_valid then
        log("✅ ライセンス認証成功！")
        return true, response.expires_at
    else
        log("❌ ライセンス認証失敗")
        return false
    end
end

-- デバイスハッシュ生成テスト
local function testDeviceHashGeneration()
    log("=== デバイスハッシュ生成テスト ===")

    -- シミュレートされたデバイス情報
    local deviceId = "test_device_001"
    local model = "iPhone"

    local data = deviceId .. ":" .. model .. ":socialtouch"
    log("ハッシュ元データ: " .. data)

    -- 簡易ハッシュ生成
    local hash = ""
    local sum = 0

    for i = 1, string.len(data) do
        sum = sum + string.byte(data, i)
    end

    math.randomseed(sum)
    for i = 1, 16 do
        local n = math.random(0, 15)
        if n < 10 then
            hash = hash .. tostring(n)
        else
            hash = hash .. string.char(87 + n)
        end
    end

    log("生成されたハッシュ: " .. hash)
    return hash
end

-- メイン実行
local function main()
    log("==========================================")
    log("    SocialTouch License Test Script      ")
    log("==========================================")

    -- デバイスハッシュ生成テスト
    local generatedHash = testDeviceHashGeneration()

    print() -- 空行

    -- ライセンス認証テスト
    local isValid, expiresAt = testLicenseVerification()

    print() -- 空行

    if isValid then
        log("🎉 すべてのテストが正常に完了しました")
        log("📅 有効期限: " .. tostring(expiresAt))

        -- Luaスクリプトでの処理例
        log("📱 実際のAutoTouchスクリプトでは以下の処理を実行:")
        log("  1. ツール選択画面を表示")
        log("  2. 選択されたツールを実行")
        log("  3. ライセンス認証は24時間キャッシュされます")
    else
        log("⚠️ ライセンス認証に失敗しました")
        log("📝 実際のスクリプトでは登録画面を表示します")
    end

    log("==========================================")
end

-- テスト実行
main()